### PR TITLE
[ENH] Added tests for MultiplexForecaster

### PR DESCRIPTION
**Pull request by miraep8**
#### Reference Issues/PRs

As mentioned in #2514 wrote some tests to see if `MultiplexForecaster` performs as expected.  

#### What does this implement/fix? Explain your changes.

Essentially - run `ForecastingGridSearchCV`  on `MultiplexForecaster` and verify that the best forecaster from the grid search is also the one we select if we simply evaluated all the forecasters separately. 

*Note - I use a very small data set and only compare have 2 forecasters in `MultiplexForecaster`.  I did this to not make the test too slow - but it is not the most thorough test.  Hopefully it would still help catch some obvious issues.*

#### Does your contribution introduce a new dependency? If yes, which one?

No

#### What should a reviewer concentrate their feedback on?

Any other obvious tests that should be added? Should we add more forecasters?  Does this test make sense to you? 

#### Any other comments?

@aiwalter I went ahead and added you as an author to this file.  I figured since you had a hand in co-creating `MultiplexForecaster` in the first place and also weighed in on #2514 it seemed reasonable.  However - please let me know if you would prefer if I removed you. 

#### PR checklist
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [x] I've added unit tests and made sure they pass locally.
- [x] The PR title starts with either [ENH], [DOC] or [BUG] indicating whether the PR topic is related to enhancement, documentation or bug

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks for this, @miraep8! This was indeed missing!

I would suggest is to add another, more confined check for consistency.
Because your test is already testing consistency of the multiplexer, integrated with grid search, i.e., not just of the multiplexer.

The more elementary consistency check would be: to check whether the `MultiplexForecaster` behaves exactly as the selected forecaster. I.e., calling `fit` then `predict` etc on one or more scenarios, on the `MultiplexForecaster`,  and on the wrapped forecaster, and comparing results to be equal. Caveats are randomness, i.e., a deterministic forecaster should be chosen or a random seed set.

**Comment by fkiraly** [comment]
& I´m looking forward to use these tests to check correctness of https://github.com/alan-turing-institute/sktime/pull/2458.

I somehow overlooked there were no estimator specific tests for the multiplex forecaster, good catch!

**Comment by miraep8** [comment]
Done! Since it seems there is some work currently being done on adding `random_state` to random forecasters like `AutoETS` I went ahead and limited myself to testing with `ThetaForecaster` and the `NaiveForecaster`.

File: sktime/forecasting/compose/tests/test_multiplex.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3 -u
+# -*- coding: utf-8 -*-
+# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+"""Tests for MultiplexForecaster and associated dunders."""
+
+__author__ = ["aiwalter", "miraep8"]
```
Comment (fkiraly) [inline sktime/forecasting/compose/tests/test_multiplex.py:6]: why is aiwalter an author? Did you take code from elsewhere?

File: sktime/forecasting/compose/tests/test_multiplex.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+from sktime.datasets import load_shampoo_sales
+from sktime.forecasting.all import (
```
Comment (fkiraly) [inline sktime/forecasting/compose/tests/test_multiplex.py:9]: better not use `forecasting.all`, it has not been used in a while, and who knows what will happen with it.

In case you want to help fix the `all` import, have a look at this issue: https://github.com/alan-turing-institute/sktime/issues/1928 (until then I would recommend to not use it)

File: sktime/forecasting/compose/tests/test_multiplex.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    """
+    from numpy.testing import assert_array_equal
+
+    y = load_shampoo_sales()
+    # Note - we select two forecasters which are deterministic.
+    forecaster_tuples = [
+        ("naive", NaiveForecaster()),
+        ("theta", ThetaForecaster()),
+    ]
+    forecaster_names = [name for name, _ in forecaster_tuples]
```
Comment (fkiraly) [inline sktime/forecasting/compose/tests/test_multiplex.py:51]: I think we need to be careful, there is a subtle point here:
objects are being passed by reference in python. So, both the `MultiplexForecaster` hyper-parameters and the `forecasters` list are pointing to the *identical* forecaster objects!
They should be cloned inside the `MultiplexForecaster`, but you could think up a situation where two bugs cancel each other out - i.e., if clone is not done properly.
I would recommend a deepcopy to have the two sides of the comparison, or separate construction from the class.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice, thanks!

One change request, related to ensuring that the two sides of the comparison or different objects and not references to the same object.

(kindly let me know if you think my reasoning is wrong here)

File: sktime/forecasting/compose/tests/test_multiplex.py (hunk 3 above)
Comment (miraep8) [reply sktime/forecasting/compose/tests/test_multiplex.py:51]: Oof, my bad!  Very important point! Will be sure to fix that immediately, thanks for spotting!

File: sktime/forecasting/compose/tests/test_multiplex.py (hunk 2 above)
Comment (miraep8) [reply sktime/forecasting/compose/tests/test_multiplex.py:9]: Sure, will fix this to be more specific with the imports!  Thanks for the heads up :)

Comment (miraep8) [reply sktime/forecasting/compose/tests/test_multiplex.py:9]: I actually think I may have copied that from the MutliplexForecaster example haha, so will be sure to go and fix it there as well. :)

File: sktime/forecasting/compose/tests/test_multiplex.py (hunk 1 above)
Comment (miraep8) [reply sktime/forecasting/compose/tests/test_multiplex.py:6]: Sorry - chalk this up to me being new to collaborative coding/open source projects!  I wanted to ere on the side of giving as much credit as possible (as I mentioned in the **Any other comments** section above!) But no - the code including the unfortunate reference error is mine! (though I certainly did look at the example `MultiplexForecaster` code as well as other test scripts for inspiration!).

File: sktime/forecasting/compose/tests/test_multiplex.py (hunk 3 above)
Comment (miraep8) [reply sktime/forecasting/compose/tests/test_multiplex.py:51]: Oh actually - I do remember this! I think I was being a bit lazy as I noticed that MultiplexForecaster does technically make a deep copy via sklearn.base.clone (see below)  But still, this code is ugly regardless and I will fix it. 

  ```
  def _set_forecaster(self):
        self._check_selected_forecaster()
        if self.selected_forecaster is not None:
            for name, forecaster in self.forecasters:
                if self.selected_forecaster == name:
                    self.forecaster_ = clone(forecaster)
```

Comment (fkiraly) [reply sktime/forecasting/compose/tests/test_multiplex.py:51]: > as I noticed that MultiplexForecaster does technically make a deep copy via sklearn.base.clone (see below)

Yes, I know - my point being, if that gets broken somehow (and it is not difficult to break this, with all the underscores or not-underscores floating around), then the test will no longer pick up if the tested property fails...

File: sktime/forecasting/compose/tests/test_multiplex.py (hunk 2 above)
Comment (miraep8) [reply sktime/forecasting/compose/tests/test_multiplex.py]: Actually - if it all right with you I might actually fix the MultiplexForecaster input example in my upcoming `|` dunder PR.  As it stands it is all mixed up with those changes I was working on but didn't necessarily want to be a part of this PR.

File: sktime/forecasting/compose/tests/test_multiplex.py (hunk 3 above)
Comment (miraep8) [reply sktime/forecasting/compose/tests/test_multiplex.py:53]: Ah, got it!  Sorry, and yes I totally agree.  Better to have code that is easier to break.  Thanks! Hopefully this resolves it for you!

File: sktime/forecasting/compose/tests/test_multiplex.py (hunk 1 above)
Comment (fkiraly) [reply sktime/forecasting/compose/tests/test_multiplex.py]: so, unless @aiwalter objects, I´d just suggest to remove it.

This is not like academia where you have to put the department head, their cousin and their dog on the paper 😄 

Also, whoever is in the author´s list is the person to blame for bugs in the file 😁

File: sktime/forecasting/compose/tests/test_multiplex.py (hunk 2 above)
Comment (fkiraly) [reply sktime/forecasting/compose/tests/test_multiplex.py]: sure, makes sense - generally it´s a good idea to keep PR localized (by file or by content)

File: sktime/forecasting/compose/tests/test_multiplex.py (hunk 1 above)
Comment (miraep8) [reply sktime/forecasting/compose/tests/test_multiplex.py]: It has been removed!   Blame me for the bugs!! 🐛 🐛 🐛

**Review by fkiraly — APPROVED** [review]
Thanks!

Looks good and tests seem to pass.
